agent/internal/detect.go

```diff
@@ -71,11 +73,69 @@ func detectCPUs() ([]device.Device, error) {
 	}
 }
 
+var detectMIGEnabled = []string{
+							"nvidia-smi", "--query-gpu=mig.mode.current", "--format=csv,noheader"}
+var detectNvidiaDevices = []string{"nvidia-smi", "-L"} // Lists both GPUs and MIG instances
+var detectMIGRegExp = regexp.MustCompile(`(?P<dev>MIG \S+).+\(UUID.+(?P<uuid>MIG.+)\)`)
+
 var detectGPUsArgs = []string{"nvidia-smi", "--query-gpu=index,name,uuid", "--format=csv,noheader"}
 var detectGPUsIDFlagTpl = "--id=%v"
 
+// detect if MIG is enabled and if there are instances configured.
+func detectMigInstances(visibleGPUs string) ([]device.Device, error) {
+	// Fail fast if MIG isn't even enabled
+	// #nosec G204
+	cmd := exec.Command(detectMIGEnabled[0], detectMIGEnabled[1:]...)
+	out, err := cmd.Output()
+	if execError, ok := err.(*exec.Error); ok && execError.Err == exec.ErrNotFound {
+		return nil, nil
+	} else if err != nil {
+		log.WithError(err).WithField("output", string(out)).Warnf(
+			"error while executing nvidia-smi to detect MIG mode")
+		return nil, nil
+	}
+	if !strings.HasPrefix(string(out), "Enabled") {
+		return nil, nil
+	}
+
+	// #nosec G204
+	cmd = exec.Command(detectNvidiaDevices[0], detectNvidiaDevices[1:]...)
+	out, err = cmd.Output()
+	if err != nil {
+		log.WithError(err).WithField("output", string(out)).Warnf(
+			"error while executing nvidia-smi to detect MIG instances")
+		return nil, nil
+	}
+
+	devices := make([]device.Device, 0)
+	deviceIndex := 0
+
+	scanner := bufio.NewScanner(strings.NewReader(string(out)))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if detectMIGRegExp.MatchString(line) {
+			matches := detectMIGRegExp.FindStringSubmatch(line)
+			if len(matches) != 3 {
+				continue
+			}
+			brand := matches[1]
+			uuid := matches[2]
+			devices = append(devices,
+				device.Device{ID: deviceIndex, Brand: brand, UUID: uuid, Type: device.GPU})
+			deviceIndex++
+		}
+	}
+	return devices, err
+}
+
 // detectGPUs returns the list of available Nvidia GPUs.
 func detectGPUs(visibleGPUs string) ([]device.Device, error) {
+	devices, err := detectMigInstances(visibleGPUs)
+	if err == nil && devices != nil && len(devices) > 0 {
```

Comment:
if we err when we should truly be detecting mig devices and fallthrough and detect normal devices, what's the behavior (of running on a full device with mig enabled)? same?

Reply:
It depends a little bit on how far you get. I believe you can use the physical GPU like normal even if MIG is enabled. But if MIG instances exist and you don't try using them, I'm actually not sure what happens. So we should definitely at least log the error here, and as a semi-separate issue I like the idea we discussed of rejecting agents if something about what they're reporting is inconsistent with the rest of the resource pool.

Reply:
Interesting - i was testing on my local (home) with a nvidia card. I'll play with mig then. Ok well if it will still work, i just won't be exactly what the user expects, and it usually won't happen unless the format of nvidia-smi changes (i hope they know the whole world is just parsing text), it's fine.

Reply:
nit, `devices != nil && len(devices) > 0` => `len(devices) > 0`